android-pdf-viewer/src/main/java/com/github/barteksc/pdfviewer/annotation/core/pdf/PdfUtil.kt

```diff
@@ -0,0 +1,327 @@
+                    documentations.add(
+                        Documentation(
+                            schemaId.intValue().toLong(),
+                            documentId.toString()
+                        )
+                    )
+                }
+            }
+        } else {
+            return null
+        }
+        return Relations(documentations)
+    }
+
+    /** Map annotations to shapes,
+     *  using the page height when converting between PDF space and image space */
+    private fun getShapesFor(
+        pdfAnnotations: List<Annotation>,
+        pageHeight: Int
+    ): List<Shape> {
+        // convert annotation to shape
+        val shapes = pdfAnnotations.map { annotation ->
+            when (annotation.type) {
+                AnnotationType.SQUARE.name -> return@map annotation.toRectangleShape(pageHeight)
+                else -> throw Exception("Annotation is not recognised")
+            }
+        }
+        return shapes
+    }
+
+    /** Use the passed PDF file path to map the PDF page to PNG image and
+     *  map PDF annotations to image shapes, save the PNG image to the given output directory
+     */
+    @JvmStatic
+    @Throws(IOException::class)
+    private fun convertPdfAnnotationsToPngShapes(
+        pdfPath: String, outputDirectory: String
+    ): PdfToImageResultData {
+        lateinit var pngFile: File
+        var pageHeight by Delegates.notNull<Int>()
+        lateinit var jsonShapes: String
+
+        val renderer = PdfRenderer(getSeekableFileDescriptor(pdfPath))
+        renderer.use { renderer ->
+            // Assuming the pdf will have only 1 page (for now)
+            val pageNum = 0
+            val page = renderer.openPage(pageNum)
+            pageHeight = page.height
+            val bitmap = Bitmap.createBitmap(page.width, page.height, Bitmap.Config.ARGB_8888)
+            // Ensure white background
+            bitmap.eraseColor(Color.WHITE)
+            page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
+
+            val pdfName = extractFileNameFromPath(pdfPath)
+            pngFile = saveBitmapAsPng(
+                bitmap,
+                outputDirectory,
+                "PdfToImage-$pdfName-page-${pageNum + 1}.png"
+            )
+
+            val pdfAnnotations = getAnnotationsFrom(pdfPath, pageNum = pageNum + 1)
+            val shapes = getShapesFor(pdfAnnotations, page.height)
+            jsonShapes = mapPdfShapesToJsonString(shapes as List<Rectangle>)
+            page.close()
+        }
+        return PdfToImageResultData(File(pdfPath), pngFile, pageHeight, jsonShapes)
+    }
+
+    private fun getSeekableFileDescriptor(pdfPath: String): ParcelFileDescriptor {
+        var fd: ParcelFileDescriptor? = null
+        try {
+            fd = ParcelFileDescriptor.open(File(pdfPath), ParcelFileDescriptor.MODE_READ_ONLY)
+        } catch (e: FileNotFoundException) {
+            e.printStackTrace()
+        }
+        return fd!!
```

Comment:
Intead of using `!!` here return null and handle that on the calling site

Reply:
Handled